src/main/scala/lib/KSA.scala

```diff
@@ -0,0 +1,56 @@
+package lib
+
+import exercise2.{Queue, QueueLike, StackEmpty}
+
+import scala.annotation.tailrec
+import scala.util.Random
+
+object KSA {
+  val ENERGY_DECAY_FACTOR = 0.996
+  val r = new Random()
+
+  /**
+   * Returns a queue containing a total of frequency elements of random values between .5 and -.5 multiplied by volume.
+   * Frequency must be greater than zero and volume is between 0 and 1.
+   */
+  def whiteNoise(frequency: Int = 440, volume: Double = 1.0): QueueLike[Double] = {
+    if(frequency <= 0 || volume < 0 || volume > 1) {
+      throw new Exception("wrong input")
+    }
+
+    // Keep adding queue until it's enough
+    @tailrec
+    def addQueue(queue: QueueLike[Double], number: Int): QueueLike[Double] = {
+      if(number == 0) return queue
+
+      val newQueue = queue.enqueue(Random.between(-0.5, 0.5) * volume)
+      addQueue(newQueue, number - 1)
+    }
+
+    addQueue(new Queue[Double](StackEmpty(), StackEmpty()), frequency)
+  }
+
+  /**
+   * Takes a queue and returns a new queue.
+   * Remove the front double in the queue and average it with the next double multiplied by newDouble.
+   * Add newDouble to the queue.
+   */
+  def update(queue: QueueLike[Double]): QueueLike[Double] = {
+    val newQueue = queue.dequeue().get
```

Comment:
`.get` on `Option` should be avoided. would be nicer if `update` itself would return `Option[QueueLike[Double]]` and then in the end, when really needed, call `.get` on that `Option`. (there are ways to avoid called that get as well using IO, which we will check out in this week's lecture)